cpp/src/interpolate/cubic_spline.cu

```diff
@@ -0,0 +1,404 @@
+  }
+};
+
+struct interpolate {
+  template <typename T>
+  std::enable_if_t<std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& ids,
+                  cudf::column_view const& coef_indexes,
+                  cudf::table_view const& coefficients,
+                  cudf::mutable_column_view const& result,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      const T* T_ = t.data<T>();
+      const int32_t* IDS_ = ids.data<int32_t>();
+      const int32_t* COEF_INDEXES_ = coef_indexes.data<int32_t>();
+      const T* D3_ = coefficients.column(3).data<T>();
+      const T* D2_ = coefficients.column(2).data<T>();
+      const T* D1_ = coefficients.column(1).data<T>();
+      const T* D0_ = coefficients.column(0).data<T>();
+      T* RESULT_ = result.data<T>();
+      thrust::for_each(rmm::exec_policy(stream)->on(stream),
+        thrust::make_counting_iterator<int>(0),
+        thrust::make_counting_iterator<int>(t.size()),
+        [T_, IDS_, COEF_INDEXES_, D3_, D2_, D1_, D0_, RESULT_] __device__
+        (int index) {
+          int h = RESULT_[index];
+          RESULT_[index] = D3_[h] + T_[index] * (D2_[h] + T_[index] * (D1_[h] + (T_[index] * D0_[h])));
+      });
+  };
+  template <typename T>
+  std::enable_if_t<not std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& ids,
+                  cudf::column_view const& coef_indexes,
+                  cudf::table_view const& coefficients,
+                  cudf::mutable_column_view const& result,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      CUDF_FAIL("Non-floating point operation is not supported.");
+  }
+};
+
+struct coefficients_compute {
+  template <typename T>
+  std::enable_if_t<std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& y,
+                  cudf::column_view const& prefixes,
+                  cudf::mutable_column_view const& h,
+                  cudf::mutable_column_view const& i,
+                  cudf::mutable_column_view const& z,
+                  cudf::mutable_column_view const& d3,
+                  cudf::mutable_column_view const& d2,
+                  cudf::mutable_column_view const& d1,
+                  cudf::mutable_column_view const& d0,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      const T* T_ = t.data<T>();
+      const T* Y_ = y.data<T>();
+      const int32_t* PREFIXES = prefixes.data<int32_t>();
+      T* H_ = h.data<T>();
+      T* I_ = i.data<T>();
+      T* Z_ = z.data<T>();
+      T* D3_ = d3.data<T>();
+      T* D2_ = d2.data<T>();
+      T* D1_ = d1.data<T>();
+      T* D0_ = d0.data<T>();
+      thrust::for_each(rmm::exec_policy(stream)->on(stream),
```

Comment:
I think this should be a thrust::transform. Since you create 4 coefficients, make the coefficients an array of float4 rather than 4 different arrays. You'll get far better bandwidth utilization on GPUs.

Reply:
Ignore the float4 suggestion. But you could use a thrust zip_iterator to turn them into 4-tuples which would have the same bandwidth benefit. Then this could be a thrust::transform() with the zip iterator as the output. Cleaner and probably faster.